contracts/utils/RolesBase.sol

```diff
@@ -0,0 +1,282 @@
+        }
+
+        _setRoles(account, accountRoles);
+
+        emit RolesAdded(account, roles);
+    }
+
+    /**
+     * @notice Internal function to remove roles from an account.
+     * @dev emits a RolesRemoved event.
+     * @param account The address to remove roles from
+     * @param roles The roles to remove
+     */
+    function _removeRoles(address account, uint8[] memory roles) internal {
+        uint256 accountRoles = _getRoles(account);
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            accountRoles &= ~(1 << roles[i]);
+        }
+
+        _setRoles(account, accountRoles);
+
+        emit RolesRemoved(account, roles);
+    }
+
+    /**
+     * @notice Internal function to check if an account has a role.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param role The role to check
+     * @return True if the account has the role, false otherwise
+     */
+    function _hasRole(uint256 accountRoles, uint8 role) internal pure returns (bool) {
+        return accountRoles & (1 << role) != 0;
+    }
+
+    /**
+     * @notice Internal function to check if an account has all the roles.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param roles The roles to check
+     * @return True if the account has all the roles, false otherwise
+     */
+    function _hasAllTheRoles(uint256 accountRoles, uint8[] memory roles) internal pure returns (bool) {
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            if (accountRoles & (1 << roles[i]) == 0) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * @notice Internal function to check if an account has any of the roles.
+     * @param accountRoles The roles of the account in uint256 format
+     * @param roles The roles to check
+     * @return True if the account has any of the roles, false otherwise
+     */
+    function _hasAnyOfRoles(uint256 accountRoles, uint8[] memory roles) internal pure returns (bool) {
+        uint256 length = roles.length;
+
+        for (uint256 i = 0; i < length; ++i) {
+            if (accountRoles & (1 << roles[i]) != 0) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @notice Internal function to propose to transfer roles of message sender to a new account.
+     * @dev Can only be called by the account with all the proposed roles.
+     * @dev emits a RolesProposed event.
+     * @dev Roles are not transferred until the new role accepts the role transfer.
+     * @param toAccount The address to transfer role to
+     * @param roles The roles to transfer
+     */
+    function _proposeRoles(address toAccount, uint8[] memory roles) internal {
+        uint256 proposed = _toAccountRoles(roles);
+
+        _setProposedRoles(msg.sender, toAccount, proposed);
```

Comment:
don't see the check that msg.sender has the `proposed` roles already to allow proposing it for another account

Reply:
it's implemented in the external function.

Reply:
ok, doc is misleading,
https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/pull/99/files#r1362575370
Can you update it for other internal functions as well?

Reply:
Maybe we actually need to do a check inside of that internal function, since transfer roles always implies fromAccount should have those roles.